src/Oracle.sol

```diff
@@ -0,0 +1,57 @@
+pragma solidity ^0.8.4;
+//SPDX-License-Identifier: MIT
+
+abstract contract Oracle {
+    error CardinalityNotAllowed();
+
+    struct Observation {
+        uint256 timestamp;
+        uint256 cumulativeWorth;
+    }
+
+    struct ObservationIndex {
+        uint128 cardinality;
+        uint128 lastIndex;
+    }
+
+    mapping(uint256 => Observation[65536]) observations;
+    mapping(uint256 => ObservationIndex) observationIndex;
+
+
+    function write(uint256 cloneId, uint256 price) internal {
+        ObservationIndex memory index = observationIndex[cloneId];
+        Observation memory lastObservation = observations[cloneId][index.lastIndex];
+        if (block.timestamp == lastObservation.timestamp) return;
+
+        unchecked {
+            if (++index.lastIndex == index.cardinality) {
+                if (observations[cloneId][index.lastIndex].timestamp != 0) {
```

Comment:
> array length is 1 greater than the limit. why? see the next comment

this saves checking that index.cardinality==65535 here. observations[cloneId][65535].timestamp's timestamp is always 0.